src/sensors/Bno055.cpp

```diff
@@ -0,0 +1,753 @@
+#include "Bno055.h"
+
+namespace rs
+{
+    /**
+     * Initialize the Bno to a default configuration state.
+     *
+     * @note This function can potentially take 1-2 seconds to execute because
+     *       the Bno055 requires 650ms to wake up from a reset and the built in
+     *       self test is given 500ms.
+     *
+     * @param use_external_crystal Specified true if the Bno should be
+     *        configured to use an external crystal.
+     *
+     * @return Zero upon success and non-zero upon failure.
+     */
+    int Bno055::init(bool use_external_crystal)
+    {
+        uint8_t self_test_result, chip_id, clock_status, clock_config;
+        AbortIf(set_page(0));
+
+        /*
+         * Perform a software reset to start off in a default state.
+         */
+        AbortIf(reset());
+
+        /*
+         * Sanity check that this is the chip we expect it to be. Result
+         * derived from [4.3.1].
+         */
+        AbortIf(read_register(Bno055::Register::CHIP_ID, chip_id));
+        AbortIfNot(chip_id == Bno055::chip_id, -1);
+
+        /*
+         * Validate that the power on self-test passed for all sensors.
+         */
+        AbortIf(read_register(Bno055::Register::ST_RESULT, self_test_result));
+        AbortIfNot(self_test_result  == 0x0F, -1);
+
+        /*
+         * Begin a built-in self-test of the device and wait for it to
+         * complete.
+         */
+        uint8_t sys_trigger_reg;
+        AbortIf(read_register(Bno055::Register::SYS_TRIGGER, sys_trigger_reg));
+        AbortIf(write_register(Bno055::Register::SYS_TRIGGER,
+                sys_trigger_reg | 1));
+        usleep(500000);
+        AbortIf(read_register(Bno055::Register::ST_RESULT,
+                self_test_result));
+        AbortIfNot(self_test_result == 0x0F, -1);
+
+        /*
+         * Read and configure the clock to utilize the external oscillator if
+         * the clock is able to be configured.
+         */
+        AbortIf(read_register(Bno055::Register::SYS_CLK_STATUS, clock_status));
+        if (use_external_crystal && (clock_status & 0b1))
+        {
+            AbortIf(write_register(Bno055::Register::SYS_TRIGGER, 1<<7));
+        }
+
+        /*
+         * Set the power and operating mode to configuration defaults.
+         */
+        AbortIf(setPowerMode(Bno055::PowerMode::Normal));
+        AbortIf(setOperationMode(Bno055::OperationMode::Config));
+
+        return 0;
+    }
+
+    /**
+     * Sets the current power setting of the sensor.
+     *
+     * @param mode The power mode to set the sensor to.
+     *
+     * @return Zero upon success or a non-zero error code returned by the
+     *         sensor.
+     */
+    int Bno055::setPowerMode(PowerMode mode)
```

Comment:
Is there a need to allow the user set the power mode? Just default to normal mode.